### PR TITLE
Feature demostart

### PR DESCRIPTION
**Pull request by RichardSmithONS**
This is the refactored demoui, containing just the home page and single match input and very little of the old code. It is not yet calling the API. Bootstrap has gone. The GDS css and scripts have been added in at the base, and some CSS is overriden to make it look like the wireframes (for now). Messages are externalized and as a test Welsh is implemeneted ?lang=cy on the query string (may have to referesh twice). The messagesAPI injection caused problems with the tests and Alex helped me fix this.

### CODE REVIEW

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
+@()(implicit messages: Messages)
```
Comment (Mironor) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html]: This still works? We moved MessagesApi from the implicit parameter to the explicit one in the Controller.

File: build.sbt [hunk 2]
```diff
@@ -56,7 +56,8 @@ lazy val localCommonSettings: Seq[Def.Setting[_]] = Seq(
     "elasticsearch-releases"     at "https://maven.elasticsearch.org/releases",
     "Java.net Maven2 Repository" at "http://download.java.net/maven/2/",
     "Twitter Repository"         at "http://maven.twttr.com",
-    "Artima Maven Repository"    at "http://repo.artima.com/releases"
+    "Artima Maven Repository"    at "http://repo.artima.com/releases",
+    "scalaz-bintray" at "https://dl.bintray.com/scalaz/releases"
```
Comment (Mironor) [inline build.sbt]: Align `at`

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import play.api.test.{FakeRequest, WithApplication}
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class ApplicationHomeTest extends PlaySpec with Results {
+  "Application controller" should {
+    "include at least one link" in new WithApplication {
+      // Given
+      val messagesApi = app.injector.instanceOf[MessagesApi]
+      val expectedString = "Find an address"
+      // When
```
Comment (Mironor) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala]: Add new line

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      val content = contentAsString(response)
+
+      // Then
+      status(response) mustBe OK
+      content must include(expectedString)
+    }
+  }
+
+
+  "Single match controller" should {
```
Comment (Mironor) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala]: You can write both specs (the `"..." in {}`) inside the same `should`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHome.scala [hunk 5]
```diff
@@ -5,12 +5,18 @@ import play.Logger
...
 @Singleton
-class ApplicationHome @Inject()(implicit ec : ExecutionContext) extends Controller {
+class ApplicationHome @Inject()(val messagesApi: MessagesApi)(implicit ec : ExecutionContext) extends Controller with I18nSupport {
 
-  def indexPage() : Action[AnyContent] = Action { implicit req =>
-    Logger.info("Rendering Index page")
-    Ok
+    def indexPage() : Action[AnyContent] = Action { implicit req =>
+      Logger.info("Rendering Index page")
+      req.getQueryString("lang") match{
```
Comment (alexflav23) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHome.scala]: @RichardSmithONS Also please put the correct indentation in code, do not skip spaces. ` req.getQueryString("lang") match {`

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala (hunk 3 above)
Comment (alexflav23) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/ApplicationHomeTest.scala]: @Mironor I'm not up to speed with the Guice injection piece, but do we get to choose between using the Scala implicit API versus the dependency injection framework?

**Comment by RichardSmithONS** [comment]
Notes on latest commit "pr issues fixed and tweak layout":

The messages are available from implicit Messages object and @inputText requires this so I haven't changed the views to use the explicit MessagesAPI which could be used instead.

The default language is obtained using the standard Play Configuration object - todo to upgrade to pure config.

I have changed the layout to more closely match  the wireframes as directed, but not fixed the overly large GDS footer (yet, so this is also a todo)

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html (hunk 1 above)
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html]: implicits can be implicit or explicitly provided

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+@import helper._
+@*
+ * Call the `main` template with two arguments. The first
+ * argument is a `String` with the title of the page, the second
+ * argument is an `Html` object containing the body of the page.
+ *@
+
+@main(title = "ONS-Address") {
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/index.scala.html]: should be  Title = Messages("ons.title")

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+
+            <div class="footer-meta">
+                <div class="footer-meta-inner">
+
+
+                    <div class="open-government-licence">
+                        <p class="logo"><a href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/" rel="license">Open Government Licence</a></p>
+
+                        <p>All content is available under the <a href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/" rel="license">Open Government Licence v3.0</a>, except where otherwise stated</p>
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html]: Messages????

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    <body>
+    <script>document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');</script>
+    <nav class="navbar navbar-default">
+        <div class="container-fluid">
+            <div class="row">
+                <div class="col-md-2 logo">
+                    <img alt="ONS" src="/assets/images/ons-logo.png" width="204" height="auto">
+                </div>
+                <div class="col-md-10 beta-header">
+                    <p><strong class="beta-banner">BETA</strong></p>
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html]: messages

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/navigation.scala.html [hunk 9]
```diff
@@ -0,0 +1,9 @@
+@()(implicit messages: Messages)
+<div>
+    <nav class="nav-container">
+        <ul>
+            <li><a href="/">@messages("navbar.home")</a></li>
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/navigation.scala.html]: use reverse routes!

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/singleMatch.scala.html [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+@import helper._
+
+@*
+ * Call the `main` template with two arguments. The first
+ * argument is a `String` with the title of the page, the second
+ * argument is an `Html` object containing the body of the page.
+ *@
+
+@main(title = "ONS-Address - Single Search") {
+<div class="container">
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/singleMatch.scala.html]: this container structure is copied everywhere, why not make it a reuasble view?

File: demo-ui/public/javascripts/ie.js [hunk 11]
```diff
@@ -0,0 +1,494 @@
+/**
+* @preserve HTML5 Shiv 3.7.3 | @afarkas @jdalton @jon_neal @rem | MIT/GPL2 Licensed
```
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/ie.js]: what is this used for?

File: demo-ui/public/javascripts/govuk-template.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
+(function () {
+  "use strict"
```
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/govuk-template.js]: what is this used for?

**Comment by RichardSmithONS** [comment]
Thanks Rhys. I'll apply the messages, reusable view and routing changes to my current open branch. I haven't studied the details of the GDS javascript as yet (it made Codacy squeal), we may not want to keep all of it.

File: demo-ui/public/javascripts/ie.js (hunk 11 above)
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/ie.js]: is this not available on webjar? We shouldnt be copy pasting lib code into our project

File: demo-ui/conf/application.conf [hunk 13]
```diff
@@ -0,0 +1,4 @@
+play.http.errorHandler = "uk.gov.ons.addressIndex.demoui.ErrorHandler"
+play.i18n.langs = [ "en", "cy" ]
+play.crypto.secret = "antidisestablishmentarianism"
```
Comment (iRhysBradbury) [inline demo-ui/conf/application.conf]: nice

File: demo-ui/conf/routes [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# Routes
+# Definition of all applications routes
+
```
Comment (iRhysBradbury) [inline demo-ui/conf/routes]: align this file

File: demo-ui/conf/routes [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+# ONS-Address Index home page
+GET     /                                       uk.gov.ons.addressIndex.demoui.controllers.ApplicationHome.indexPage
+
+# ONS-Address Index Single Search View
+GET     /addresses                              uk.gov.ons.addressIndex.demoui.controllers.SingleMatch.showSingleMatchPage
+
+# ONS-Address Index Multiple Search
+#GET     /addresses/bulk                         uk.gov.ons.addressIndex.demoui.controllers.MultipleMatch.showMultipleMatchPage
+
```
Comment (iRhysBradbury) [inline demo-ui/conf/routes]: lots of commented out routes?

File: demo-ui/public/javascripts/govuk-template.js [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+      if(value === false || value === null) {
+        return GOVUK.setCookie(name, '', { days: -1 });
+      } else {
+        return GOVUK.setCookie(name, value, options);
+      }
+    } else {
+      return GOVUK.getCookie(name);
+    }
+  };
+  GOVUK.setCookie = function (name, value, options) {
```
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/govuk-template.js]: where is this used?

File: demo-ui/public/javascripts/govuk-template.js [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+      var date = new Date();
+      date.setTime(date.getTime() + (options.days * 24 * 60 * 60 * 1000));
+      cookieString = cookieString + "; expires=" + date.toGMTString();
+    }
+    if (document.location.protocol == 'https:'){
+      cookieString = cookieString + "; Secure";
+    }
+    document.cookie = cookieString;
+  };
+  GOVUK.getCookie = function (name) {
```
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/govuk-template.js]: where is this used?

File: demo-ui/public/javascripts/govuk-template.js [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    return null;
+  };
+}).call(this);
+(function () {
+  "use strict"
+  var root = this;
+  if(typeof root.GOVUK === 'undefined') { root.GOVUK = {}; }
+
+  GOVUK.addCookieMessage = function () {
```
Comment (iRhysBradbury) [inline demo-ui/public/javascripts/govuk-template.js]: where is this used?

**Pull request merged**